README
eAccelerator for PHP
=====================

What is eAccelerator?
----------------------
eAccelerator is a free open source PHP accelerator and optimizer for PHP. It
increases the performance of PHP scripts by caching them in compiled state,
so the overhead of compiling  is  almost completely eliminated. It also
optimizes the scripts to speed up execution.

eAccelerator typically reduces server load and increases the speed of your
PHP code by 1-10 times.

eAccelerator is a fork of TurckMMCache 
( http://sourceforge.net/project/turckmm-cache/  by Dmitry Stogov )

eAccelerator stores compiled PHP scripts in shared memory  and  executes  code
directly from it. It creates  locks  only  for  short  time,  while  searching 
compiled PHP script in the cache, so one script can be executed simultaneously
by several engines. Files that can't fit in shared memory are cached  on  disk
only.

eAccelerator is compatible with Zend Optimizer's loader. Zend  Optimizer  must
be installed after eAccelerator in php.ini. If you don't use  scripts  encoded
with  Zend  Encoder  we  do  not  recommend  to  install  Zend  Optimizer with 
eAccelerator.

eAccelerator does not work in CGI mode but it does work in Fast-CGI mode with
webservers like lighttpd.


Download
--------
Latest eAccelerator versions can be downloaded at the sourceforge page:
http://sourceforge.net/projects/eaccelerator/
Development snapshots from cvs can be downloaded at 
http://snapshots.eaccelerator.net


Requirements
------------
* apache >= 1.3, mod_php >= 5.1 or php with fastcgi support
* autoconf, automake, libtool, m4

Compatibility
-------------
eAccelerator works with php version 5.1 and higher.
It is being used on Linux, FreeBSD, Mac OS X, Solaris and Windows 
with Apache 1.3 and 2, lighttpd and IIS.


Quick install
-------------

You can find more information about the installation of eAccelerator on our
website at http://www.eaccelerator.net/

Note(1): for Microsoft Windows installation, please refer to README.win32 file.

Step 1. Compiling eAccelerator

  export PHP_PREFIX="/usr"
  
  $PHP_PREFIX/bin/phpize
  
  ./configure \
  --enable-eaccelerator=shared \
  --with-php-config=$PHP_PREFIX/bin/php-config
  
  make

  You must specify the real prefix where PHP is installed in the "export"
  command. It may be "/usr" "/usr/local", or something else.

Step 2. Installing eAccelerator

  make install

Step 3. Configuring eAccelerator

eAccelerator can be installed both as Zend or PHP extension.

For eaccelerator > 0.9.1, if you have /etc/php.d directory, you should copy eaccelerator.ini inside and modify default value if you need.

If not, you need to edit your php.ini file (usually /etc/php.ini).

To install as Zend extension:

  zend_extension="/usr/lib/php5/eaccelerator.so"
  eaccelerator.shm_size="16"
  eaccelerator.cache_dir="/tmp/eaccelerator"
  eaccelerator.enable="1"
  eaccelerator.optimizer="1"
  eaccelerator.debug="0"
  eaccelerator.filter=""
  eaccelerator.shm_ttl="0"
  eaccelerator.shm_prune_period="0"
  eaccelerator.shm_only="0"

  If you use thread safe build of PHP you must use "zend_extension_ts" instead
  of "zend_extension".

To install as PHP extension:

  extension="eaccelerator.so"
  eaccelerator.shm_size="16"
  eaccelerator.cache_dir="/tmp/eaccelerator"
  eaccelerator.enable="1"
  eaccelerator.optimizer="1"
  eaccelerator.debug="0"
  eaccelerator.filter=""
  eaccelerator.shm_ttl="0"
  eaccelerator.shm_prune_period="0"
  eaccelerator.shm_only="0"

Step 4. Creating cache directory

  mkdir /tmp/eaccelerator
  chmod 0777 /tmp/eaccelerator


Configuration Options
---------------------

eaccelerator.shm_size
    The amount of shared memory (in megabytes) that eAccelerator will use.
    "0" means OS default. Default value is "0".

eaccelerator.cache_dir
    The directory that is used for disk cache. eAccelerator stores precompiled
    code, session data, content and user entries  here. The same data  can  be
    stored in shared memory also (for more quick access). Default value is
    "/tmp/eaccelerator".

eaccelerator.enable
    Enables or disables eAccelerator. Should be "1" for enabling  or  "0"  for
    disabling. Default value is "1".

eaccelerator.optimizer
    Enables or disables internal peephole optimizer which may  speed  up  code
    execution. Should be "1" for enabling or "0" for disabling. Default  value
    is "1".

eaccelerator.debug
    Enables or disables debug logging. Should be "1" for enabling or  "0"  for
    disabling. Default value is "0".

eaccelerator.filter
    Determine which PHP files must be cached. You may specify  the  number  of
    patterns (for example "*.php *.phtml") which specifies to cache or not  to
    cache. If pattern starts with the character "!", it means to ignore  files
    which are matched by the following pattern. Default value is "" that means 
    all PHP scripts will be cached.

eaccelerator.shm_ttl
    When eaccelerator fails to get shared memory for new script it removes all
    scripts which were not accessed  at  last "shm_ttl"  seconds  from  shared
    memory. Default value is "0" that means -  don't  remove  any  files  from
    shared memory.

eaccelerator.shm_prune_period
    When eaccelerator fails to get shared memory for new script  it  tryes  to
    remove  old  script   if   the   previous   try   was   made   more   then
    "shm_prune_period" seconds ago. Default value is "0" that  means  -  don't
    try to remove any files from shared memory.

eaccelerator.shm_only
    Enables or disables caching of compiled scripts on disk. It has  no  effect
    on session data and content caching. Default value is "0" that means -  use
    disk and shared memory for caching.

eaccelerator.allowed_admin_path
    The script paths that are allowed to get admin information and do admin 
    controls

Control panel and disassembler
------------------------------

If you want to use the control-panel you need to compile eAccelerator with 
    --with-eaccelerator-info which is the default value.
You need to copy the control.php file to your webroot and set the path to it 
in the php.ini or eaccelerator.ini in the eaccelerator.allowed_admin_path 
directive. If you don't do this you wont be able to see much information and 
can't control eAccelerator.
You can set the username and password needed to access the control-panel in 
the control.php file. By default the username is administrator and the password
is eAccelerator.

When you compile eAccelerator with --with-eaccelerator-disassembler you need 
to place the dasm.php and PHP_Highlight.php file also in the same directory 
as the control.php file. The disassembler requires PHP to be compiled with
tokenizer support (--enable-tokenizer).
You can set the username and password needed to access the disassembler at the
top of dasm.php.

eAccelerator API
----------------

API documentation can be found on this website: http://bart.eaccelerator.net/doc/phpdoc/

Contact us
----------
You can contact us with questions, patches or bugs, by sending an email to
Bart Vanbrabant <bart at vanbrabant.eu>


ChangeLog
2010-07-11 Bart Vanbrabant <bart@vanbrabant.eu>

    * Changed the way the cache directory is created:
		- The forked process creates the cache dir and makes it world writable
		- The first request creates a subdir with the userid of the process and
		with permissions 700.

	* Comment out an unused function to support compilin with -Werror
	* Add sched.h include to fix a compiler warning

2010-06-03	Hans Rakers <hans at react.nl>

		* Fixed compilation against php-5.1. This closes ticket #426.

2010-02-25	Hans Rakers <hans at react.nl>

		* Fixed class static variable restore where parent was protected
		  static and child was public static. This would lead to error
		  'Access to undeclared static property' upon restore from cache
		* Added proper handling of class properties during storing in
		  cache. Public/private/protected is properly handled now. Somehow
		  this was a long overdue TODO spotted by running the testsuite.
		  This fixes ticket #408
		* Small fix that solves some issues with lambda (closure)
		  functions. This solves ticket #405

2010-02-24	Hans Rakers <hans at react.nl>

		* Fixed double free in eaccelerator_globals_dtor. This closes
		  ticket #404

2010-02-18	Hans Rakers <hans at react.nl>

		* Remove some leftover debug output that broke compilation with
		  php-5.2 and --with-eaccelerator-debug
		* php_check_open_basedir only needs to be called when restoring

2010-02-12  Bart Vanbrabant <bart at vanbrabant.eu>

		* Refactor the cache to make it more general

2010-02-05  Bart Vanbrabant <bart at vanbrabant.eu>
		
		* Remove more unused stuff
		* Implement ttl in new cache code
		* Implement pruning and purging the memory ht and the file cache

2010-02-04	Hans Rakers <hans at react.nl>

		* Fixed linking of ZEND_DECLARE_INHERITED_CLASS_DELAYED across
		  bb's in the optimizer. This fixes bug #399

2010-02-04	Bart Vanbrabant <bart at vanbrabant.eu>

		* First very early work to rewrite the caching code.
		* Remove some leftovers from the lock api remotes/origin/master
		* Fix a small compile warning in a printf function
		* Get rid of all EAG(mem) globals
		* Remove some leftovers from the lock api
		* Now really add the sched_yield command and not only in commented
		out

2010-02-03	Bart Vanbrabant <bart at vanbrabant.eu>

		* Include the resolve path function from php so it is available for
		older PHP versions as well. Includes that start file:// work now.
		* Enable phar caching: Files in included from a phar file are cached
		now. The phar archive is used to excute the stat on.
		* Enable check_mtime usage again
		* Use snprintf the right way. See #290
		* Add a call to sched_yield after a failed spinlock to prevent
		deadlock conditions. This should fix #381 and is a real fix for #340.
		Instead of using a fake select as yield, a check is added to disable
		spinlocks on systems that do not have sched_yield.
		* Fix spinlocks on Solaris with Sun Compiler. #325

2010-02-02  Bart Vanbrabant <bart at vanbrabant.eu>

		* Fix two compiler warnings

2010-02-01  Bart Vanbrabant <bart at vanbrabant.eu>

		* A retab on eaccelerator.c
		* Remove caching based on inodes
		* Reimplement eaccelerator_stat to speed it up: the new ea_stat is
		almost twice as fast as the old version with inodes. The tests were
		done on a file that includes 10k empty files.
		* Increment version to 1.0-dev

2010-01-22	Bart Vanbrabant <bart at vanbrabant.eu>

		* Release 0.9.6-rc2

2010-01-21	Hans Rakers <hans at react.nl>

		* Set op_array->early_binding in a later stage.
		  This fixes ticket #394

2010-01-20	Hans Rakers <hans at react.nl>

		* Some more work on delayed early binding. This fixes
		  ticket #388
		* Some reformatting in ea_dasm.c
		* Win32 VC6 project file updated

2010-01-19	Bart Vanbrabant <bart at vanbrabant.eu>

		* Remove obsolete documentation
		* Update year in copyright

2010-01-19	Hans Rakers <hans at react.nl>

		* Implemented ZEND_GOTO
		* Implemented delayed early binding for php-5.3. This
		  solves ticket #146 (Problems with require_once with
		  circular dependencies) for everyone using php-5.3.
		  See http://marc.info/?l=php-internals&m=120539866910125&w=2
		  for more info on this.
		* Fixed ticket #386 (Crash using const & php-5.3)
		* Fixed a bunch of other php-5.3 related issues found by
		  using the test suite and all php-5.3 Zend engine tests
		* Win32 project files updated
		* Win32 ZTS fixes

2009-12-23	Hans Rakers <hans at react.nl>

		* Fixed ticket #383 (crash on ZEND_ADD_INTERFACE)
		* check_mtime can now be enabled/disabled through the web
		  interface as requested in ticket #32
		* Shortened some variable names
		* ZEND_JMP_SET support

2009-12-01	Hans Rakers <hans at react.nl>

		* Win32 compilation fixes (compiles in VC++ 2008)
		* Updated VC++ project file to VC++ 2008 format

2009-11-30	Hans Rakers <hans at react.nl>

		* Added a whole bunch of standard functions to
		  opt_result_is_numeric()
		* init allocated memblocks with memset instead of a for loop
		  in reassign_registers()
		* Fix in build_cfg for PHP 5.3 UCLASS znodes. These are
		  handled differently in 5.3 due to namespaces.
		* Some tweaks and additions to ea_dasm.c
		* Some quick and dirty tweaks to dasm.php to sync looks with
		  control.php
		* Fixed ordering of table columns in control.php. This solves
		  ticket #364

2009-11-25	Hans Rakers <hans at react.nl>

		* Optimizer and general PHP 5.3 fixes
		* Some cleanups

2009-11-11	Bart Vanbrabant <bart at vanbrabant.eu>

		* Add the optimizer back

2009-11-11	Hans Rakers <hans at react.nl>

		* PHP 5.3 ZEND_JMP_SET support. Fixes ticket #371

control.php
<?php 
/*
   	+-----------------------------------------------------------------------+
   	| eAccelerator control panel                                           	|
   	+-----------------------------------------------------------------------+
   	| Copyright (c) 2004-2010 eAccelerator                                 	|
	| http://eaccelerator.net                                              	|
   	+-----------------------------------------------------------------------+
   	| This source file is subject to version 3.0 of the PHP license,       	|
	| that is bundled with this package in the file LICENSE, and is        	|
	| available through the world-wide-web at the following url:           	|
	| http://www.php.net/license/3_0.txt.                                  	|
	| If you did not receive a copy of the PHP license and are unable to   	|
	| obtain it through the world-wide-web, please send a note to          	|
	| license@php.net so we can mail you a copy immediately.             	|
	+-----------------------------------------------------------------------+
*/

/*** CONFIG ***/
$auth = true;		// Set to false to disable authentication
$user = "admin";
$pw = "eAccelerator";

$npp = 50;		// Number of records per page (script / key cache listings)

/*** TODO for API / reporting / this script: 
     + want script ttl from API
     + would be cool to have init_time for scripts (time of caching) - could then get hit rates etc.
*/

// Inline media
if (isset($_GET['img']) && $_GET['img']) {
    $img = strtolower($_GET['img']);
    $imgs['dnarr'][0] = 199;
    $imgs['dnarr'][1] = 'H4sIAAAAAAAAA3P3dLOwTORlEGBoZ2BYsP3Y0t0nlu85ueHQ2U1Hzu86efnguetHL968cPPBtbuPbzx4+vTV24+fv3768u3nr9+/f//59+/f////GUbBKBgWQPEnCzMDgyCDDogDyhMMHP4MyhwyHhsWHGzmENaKOSHAyMDAKMWTI/BAkYmDTU6oQuAhY2M7m4JLgcGDh40c7HJ8BQaBBw4z8bMaaOx4sPAsK7voDZ8GAadTzEqSXLJWBgoM1gBhknrUcgMAAA==';
    $imgs['uparr'][0] = 201;
    $imgs['uparr'][1] = 'H4sIAAAAAAAAA3P3dLOwTORlEGBoZ2BYsP3Y0t0nlu85ueHQ2U1Hzu86efnguetHL968cPPBtbuPbzx4+vTV24+fv3768u3nr9+/f//59+/f////GUbBKBgWQPEnCzMDgyCDDogDyhMMHIEMyhwyHhsWHGzmENaKOTFBoYWZgc/BYQVDw1EWdvGIOzsWJDAzinFHiBxIWNDMKMbv0sCR0NDMIcATofJB4RAzkxivg0OCoUNzIy9ThMuFDRqHGxisAZtUvS50AwAA';

    if (!$imgs[$img] || strpos($_SERVER['HTTP_ACCEPT_ENCODING'], 'gzip') === false) 
        exit();

    header("Expires: ".gmdate("D, d M Y H:i:s", time()+(86400*30))." GMT");
    header("Last-Modified: ".gmdate("D, d M Y H:i:s", time())." GMT");
    header('Content-Length: '.$imgs[$img][0]);
    header('Content-Type: image/gif');
    header('Content-Encoding: gzip');

    echo base64_decode($imgs[$img][1]);
    exit();
}

// Authenticate before proceeding
if ($auth && (!isset($_SERVER['PHP_AUTH_USER']) || !isset($_SERVER['PHP_AUTH_PW']) ||
        $_SERVER['PHP_AUTH_USER'] != $user || $_SERVER['PHP_AUTH_PW'] != $pw)) {
    header('WWW-Authenticate: Basic realm="eAccelerator control panel"');
    header('HTTP/1.0 401 Unauthorized');
    exit;
} 

$sec = isset($_GET['sec']) ? (int)$_GET['sec'] : 0;

// No-cache headers
header("Expires: Mon, 26 Jul 1997 05:00:00 GMT");
header("Last-Modified: " . gmdate("D, d M Y H:i:s") . " GMT");
header("Cache-Control: no-store, no-cache, must-revalidate");
header("Cache-Control: post-check=0, pre-check=0", false);
header("Pragma: no-cache");

function print_footer() {
    global $info;
?>
</div>
<div class="footer">
<?php
if (is_array($info)) {
?>
<br/><br/>
    <hr style="width:500px; color: #cdcdcd" noshade="noshade" size="1" />
    <strong>Created by the eAccelerator team &ndash; <a href="http://eaccelerator.net">http://eaccelerator.net</a></strong><br /><br />
    eAccelerator <?php echo $info['version']; ?> [shm:<?php echo $info['shm_type']?> sem:<?php echo $info['sem_type']; ?>]<br />
    PHP <?php echo phpversion();?> [ZE <?php echo zend_version(); ?>]<br />
    Using <?php echo php_sapi_name();?> on <?php echo php_uname(); ?><br />
<?php
}
?>
</div>
</body>
</html>
<?php
}

if (!function_exists('eaccelerator_info')) {
    die('eAccelerator isn\'t installed or isn\'t compiled with info support!');
}

// formats sizes
function format_size ($x) {
    $a = array('bytes', 'kb', 'mb', 'gb');
    $i = 0;
    while ($x >= 1024) {
        $i++;
        $x = $x / 1024;
    }
    return number_format($x, ($i > 0)?2:0, '.', ',').' '.$a[$i];
}

// Generates a simple & colourful horizontal bar graph. $x:$y is used:free
function space_graph ($x, $y) {
    $colr = 183; $colg = 225; $colb = 149;	// #B7E195

    $colr = base_convert($colr + floor(($x/$y)*(50+exp($x*3/$y))), 10, 16);
    $colg = base_convert($colg - floor(($x/$y)*(100+exp($x*4/$y))), 10, 16);
    $colb = base_convert($colb - floor(($x/$y)*(70+exp($x*4/$y))), 10, 16);

    $s = '<table class="hgraph"><tr>';
    $s .= '<td class="hgraph_pri" style="width:'.floor(($x/$y)*100).'%">&nbsp;</td>';
    $s .= '<td class="hgraph_sec" style="background-color: #'.$colr.$colg.$colb.'; width:'.ceil(100 - ($x/$y)*100).'%">&nbsp;</td></tr></table>';
    return $s;
}

// Messy algorithm to generate neat page selectors
function pageselstr ($pg, $pgs) {
    $pg += 1;
    $st = max(1, $pg - 2) - max(0, 2 - ($pgs - $pg));
    $nd = $pg + 2 + max(0, 3 - $pg);
    $d = $st - 1 - 1;
    if (abs($nd - $pg) > 2) $sp[] = $nd - 2;
    if (abs($pg - $st) > 2) $sp[] = $st + 2;
    if (($d-2)/3 >= 2) {
        $sp[] = (ceil($d/3)+1);
        $sp[] = (ceil($d/3)*2+1);
    }
    elseif (($d-1)/2 > 0) $sp[] = (ceil($d/2)+1);
    $d = $pgs - $nd - 1;
    if (($d-2)/3 >= 2) {
        $sp[] = (ceil($d/3)+$nd);
        $sp[] = (ceil($d/3)*2+$nd);
    }
    elseif (($d-1)/2 > 0) $sp[] = (ceil($d/2)+$nd);
    $sp[] = $st;$sp[] = $nd;$sp[] = $pg;$sp[] = 1;$sp[] = $pgs;
    $lp = 1;
    $pgstr = 'Page: ';
    if ($pgs < 1) $pgstr .= '-';
    for ($i = 1; $i <= $pgs; $i++) {
        if (in_array($i, $sp)) {
            if ($i - $lp <= 2 && $i > 2 && !in_array($i-1, $sp)) $pgstr .= '<a href="'.$_SERVER['PHP_SELF'].'?'.qstring_update(array('pg' => $i-2)).'">'.($i-1).'</a> ';
            if ($i - $lp > 2) $pgstr .= '..';
            if ($i == $pg) $pgstr .= ' ['.$i.'] ';
            else {
                $pgstr .= ' <a href="'.$_SERVER['PHP_SELF'].'?'.qstring_update(array('pg' => $i-1)).'">'.$i.'</a> ';
            }
            $lp = $i;
        }
    }
    return $pgstr;
}

// Returns qstring with updated key / value pairs.
function qstring_update ($arr) {
    $qs = '';
    $combo = array_merge($_GET, $arr);
    foreach ($combo as $a => $b) {
        if ($qs) $qs .= '&';
        $qs .= urlencode($a).'='.urlencode($b);
    }
    return $qs;
}

// Returns standard column headers for the lists
function colheadstr ($nme, $id) {
    $cursrt = isset($_GET['ordby']) ? $_GET['ordby'] : 0;
    $srtdir = isset($_GET['dir']) ? $_GET['dir'] : "";
    return '<a href="'.$_SERVER['PHP_SELF'].'?'.qstring_update(array('ordby' => $id, 'dir' => ($cursrt == $id)?1-$srtdir:0)).'">'.$nme.'&nbsp;'.(($cursrt == $id)?'<img src="'.$_SERVER['PHP_SELF'].'?img='.(($srtdir)?'dnarr':'uparr').'" width="13" height="16" border="0" alt="'.(($srtdir)?'v':'^').'"/>':'');
}

// Array sorting callback function
function arrsort ($a, $b) {
    global $ordby, $ordbystr;
    if ($ordbystr) 
        $val = strnatcmp($a[$ordby], $b[$ordby]);
    else 
        $val = ($a[$ordby] == $b[$ordby]) ? 0 : (($a[$ordby] < $b[$ordby]) ? -1: 1);
    if (isset($_GET['dir']) && $_GET['dir'])
        $val = -1*$val;
    return $val;
}

// Global info array
$info = eaccelerator_info();

?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">  
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
    <title>eAccelerator control panel</title>
    <meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" />
    <meta http-equiv="Content-Style-Type" content="text/css" />
    <meta http-equiv="Content-Language" content="en" />

    <style type="text/css" media="all">
        body {background-color: #ffffff; color: #000000;margin: 0px;}
        body, td {font-family: Tahoma, sans-serif;font-size: 12pt;}

        a:link {color: #ff0000; text-decoration: none;}
        a:visited {color: #ff0000; text-decoration: none;}
        a:active {color: #aa0000; text-decoration: none;}
        a:hover {color: #aa0000; text-decoration: none;}
        
        .head1 {background-color: #A9CFDD; width: 100%; font-size: 32px; color: #ffffff;padding-top: 20px;font-family: Tahoma, sans-serif;}
        .head1_item {padding-left: 15px;}
        .head2 {background-color: #62ADC2; width: 100%; font-size: 18px; color: #ffffff;text-align: right; font-family: Tahoma, sans-serif;border-top: #ffffff 2px solid;}
        .head2 a:link {color: #ffffff;}
        .head2 a:visited {color: #ffffff;}
        .head2 a:active {color: #ffffff;}
        .head2 a:hover {color: #000000;}

        .menuitem {padding-left: 15px;padding-right: 15px;}
        .menuitem_sel {padding-left: 15px;padding-right: 15px;background-color: #ffffff; color: #000000;}
        .menuitem_hov {padding-left: 15px;padding-right: 15px;cursor:pointer;color: #000000;}
        .mnbody {padding:15px; padding-top: 30px; margin-right: auto; margin-left: auto; text-align: center;bottom: 60px;}

        .mnbody table {border-collapse: collapse; margin-right: auto; margin-left: auto;}
        
        td {padding: 3px 10px 3px 10px; border: #ffffff 2px solid;vertical-align:top}
        .el {text-align: left;background-color: #e1eff8;}
        .er {text-align: right;background-color: #e1eff8;}
        .ec {text-align: center;background-color: #e1eff8;}

        .fl {text-align: left;background-color: #efefef;}
        .fr {text-align: right;background-color: #efefef;}

        .h {text-align: center; font-weight: bold;}
        .h a:link {color: #000000;}
        .h a:visited {color: #000000;}
        .h a:active {color: #ababab;}
        .h a:hover {color: #ababab;}

        .hgraph {width:100%;}
        .hgraph td {border: 0px;padding: 0px;}
        .hgraph_pri {background-color: #62ADC2;}
        .hgraph_sec {}
        
        
        .footer {width: 100%;text-align: center;font-size: 9pt;color: #ababab;}
        .footer a:link {color: #ababab;}
        .footer a:visited {color: #ababab;}
        .footer a:active {color: #000000;}
        .footer a:hover {color: #000000;}
        
        small {font-size: 10pt;}
        .s {color: #676767;}

    </style>
    
    <script type="text/javascript">
      function menusel(i) {
        if (i.className == "menuitem_hov") i.className = "menuitem";
        else if (i.className == "menuitem") i.className = "menuitem_hov";
      }
      function gosec(i) {
        document.location = "<?php echo $_SERVER['PHP_SELF']?>?sec="+i;
      }
    </script>
    
</head>

<body>
<div class="head1"><span class="head1_item">eAccelerator control panel</span></div>
<div class="head2">
<?php
$items = array(0 => 'Status', 1 => 'Script Cache');

foreach ($items as $i => $item) {
    echo '<span class="menuitem'.(($sec == $i)?'_sel':'').'" onmouseover="menusel(this)" onmouseout="menusel(this)" onclick="gosec('.$i.')">'.(($sec != $i)?'<a href="'.$_SERVER['PHP_SELF'].'?sec='.$i.'">'.$item.'</a>':$item).'</span>';
}  
?>
</div>
<div class="mnbody">
<?php
switch ($sec) {
    default:
    case 0:
        /******************************     STATUS / CONTROL     ******************************/

        if (isset($_POST['cachingoff'])) eaccelerator_caching(false);
        if (isset($_POST['cachingon'])) eaccelerator_caching(true);

        if (isset($_POST['optoff']) && function_exists('eaccelerator_optimizer')) eaccelerator_optimizer(false);
        if (isset($_POST['opton']) && function_exists('eaccelerator_optimizer')) eaccelerator_optimizer(true);

        if (isset($_POST['prune'])) eaccelerator_prune();
        if (isset($_POST['purge'])) eaccelerator_purge();

        $info = eaccelerator_info();

?>
<table>
<tr><td>

<form action="<?php echo $_SERVER['PHP_SELF']?>?sec=0" method="post">
<table>
<tr>
    <td class="h" colspan="2">Usage statistics</td>
</tr>
<tr>
    <td class="er">Caching enabled</td> 
    <td class="fl"><?php echo $info['cache'] ? '<span style="color:green"><b>yes</b></span>&nbsp;&nbsp;&nbsp;<input type="submit" name="cachingoff" value=" Disable "/>':'<span style="color:red"><b>no</b></span>&nbsp;&nbsp;&nbsp;<input type="submit" name="cachingon" value=" Enable "/>' ?></td>
</tr>
<tr>
    <td class="er">Optimizer enabled</td>
    <td class="fl"><?php echo $info['optimizer'] ? '<span style="color:green"><b>yes</b></span>&nbsp;&nbsp;&nbsp;<input type="submit" name="optoff" value=" Disable "/>':'<span style="color:red"><b>no</b></span>&nbsp;&nbsp;&nbsp;<input type="submit" name="opton" value=" Enable "/>' ?></td>
</tr>
<tr>
    <td class="er">Total memory</td>
    <td class="fl"><?php echo format_size($info['memorySize']); ?></td>
</tr>
<tr>
    <td class="er">Memory in use</td>
    <td class="fl"><?php echo format_size($info['memoryAllocated']).' ('.number_format(100 * $info['memoryAllocated'] / max(1, $info['memorySize']), 0).'%)';?></td>
</tr>
<tr>
    <td class="er" colspan="2"><?php echo space_graph($info['memoryAllocated'], $info['memorySize']);?></td>
</tr>
<tr>
    <td class="er">Free memory</td>
    <td class="fl"><?php echo format_size($info['memoryAvailable'])?></td>
</tr>
<tr>
    <td class="er">Hashtable size</td>
    <td class="fl"><?php echo number_format($info['htSize']); ?></td>
</tr>
<tr>
    <td class="er">Hashtable loadfactor</td>
    <td class="fl"><?php echo number_format(($info['cachedScripts'] / $info['htSize']) * 100, 2) ?>%</td>
</tr>
<tr>
    <td class="er">Cached scripts</td>
    <td class="fl"><?php echo number_format($info['cachedScripts']); ?></td>
</tr>
</table>
</form>

</td><td>

<table>
<tr>
    <td class="h" colspan="2">Build information</td>
</tr>
<tr>
    <td class="er">eAccelerator version</td> 
    <td class="fl"><?php echo $info['version']; ?></td>
</tr>
<tr>
    <td class="er">Shared memory type</td> 
    <td class="fl"><?php echo $info['shm_type']; ?></td>
</tr>
<tr>
    <td class="er">Semaphore type</td> 
    <td class="fl"><?php echo $info['sem_type']; ?></td>
</tr>
</table>

</td></tr>
</table>

<br/><br/>

<form action="<?php echo $_SERVER['PHP_SELF']?>?sec=0" method="post">
<table>
<tr>
    <td class="h" colspan="2">Maintenance</td>
</tr>
<tr>
    <td class="ec"><input type="submit" name="prune" value=" Prune cache "/></td> 
    <td class="fl">Prune all expired scripts and data from shared memory.</td>
</tr>
<tr>
    <td class="ec"><input type="submit" name="purge" value=" Purge cache "/></td> 
    <td class="fl">Purge all scripts and data from shared memory and / or disk.</td>

</tr>
</table>
</form>

<?php
        break;
    case 1:
        /******************************     SCRIPT CACHE     ******************************/
    
        $scripts = eaccelerator_cached_scripts();
    
        // search
        function scriptsearch ($val) {
            $str = isset($_GET['str']) ? $_GET['str'] : '';
            return preg_match('/'.preg_quote($str, '/').'/i', $val['file']);
        }
        $scripts = array_filter($scripts, 'scriptsearch');
    
        // sort
        $ordby = isset($_GET['ordby']) ? intval($_GET['ordby']) : 0;
        switch ($ordby) {
            default:
            case 0: $ordby = 'file'; $ordbystr = true; break;
            case 1: $ordby = 'mtime'; $ordbystr = false; break;
            case 2: $ordby = 'ts'; $ordbystr = false; break;
            case 3: $ordby = 'ttl'; $ordbystr = false; break;
            case 4: $ordby = 'size'; $ordbystr = false; break;
            case 6: $ordby = 'hits'; $ordbystr = false; break;
        }
        usort($scripts, 'arrsort');
 
        // slice
        $numtot = count($scripts);

        $pg = (isset($_GET['pg']) ? (int)$_GET['pg'] : 0); // zero-starting
        $pgs = ceil($numtot/$npp);

        if ($pg + 1 > $pgs)
            $pg = $pgs-1;
        if ($pg < 0)
            $pg = 0;

        $scripts = array_slice($scripts, $pg*$npp, $npp);
        $numres = count($scripts);
?>
<table class="center">
<tr>
    <td class="h" colspan="2">Search</td>
</tr>
<tr>
    <form action="<?php echo $_SERVER['PHP_SELF']?>" method="get"><input type="hidden" name="sec" value="1"/>
    <td class="el">Match filename:</td> 
    <td class="fl"><input type="text" name="str" size="40" value="<?php echo isset($_GET['str']) ? $_GET['str'] : '' ?>"/>&nbsp;<input type="submit" value=" Find "/></td>
    </form>
</tr>
</table>

<br/><br/>

<?php
        if (count($scripts) == 0) 
            echo '<div class="center"><i>No scripts found</i></div>';
        else {
?>
<table class="center">
<tr>
    <td colspan="1" style="text-align: left">Showing <?php echo $pg*$npp+1?> &ndash; <?php echo $pg*$npp+min($npp, $numres)?> of <?php echo $numtot?></td>
    <td colspan="4" style="text-align: right;"><small><?php echo pageselstr($pg, $pgs)?></small></td>
</tr>
<tr>
    <td class="h"><?php echo colheadstr('File', 0)?></td>
    <td class="h"><?php echo colheadstr('Last Modified', 1)?></td>
    <td class="h"><?php echo colheadstr('Added', 2)?></td>
    <td class="h"><?php echo colheadstr('TTL', 3)?></td>
    <td class="h"><?php echo colheadstr('Size', 4)?></td>
    <td class="h"><?php echo colheadstr('Hits', 6)?></td>
</tr>
<?php
            $disassembler = function_exists('eaccelerator_dasm_file');
            $ea_shm_ttl = ini_get("eaccelerator.shm_ttl");
            for ($i = 0; $i < $numres; $i++) {
                $removed = (isset($scripts[$i]['removed']) && $scripts[$i]['removed']);
                if ($disassembler && !$removed) {
                    $file_col = sprintf('<a href="dasm.php?file=%s">%s</a>', $scripts[$i]['file'], $scripts[$i]['file']);
                } elseif ($removed) {
                    $file_col = sprintf('<span class="s">%s</span>', $scripts[$i]['file']);   
                } else {
                    $file_col = $scripts[$i]['file'];
                }

                if ($scripts[$i]['atime'] != 0 && !empty($ea_shm_ttl)) {
                    $ttl_col = ($scripts[$i]['atime'] + $ea_shm_ttl) - time();
                    if ($ttl_col <= 0) {
                        $ttl_col = "expired";
                    }
                } else {
                    $ttl_col = "&infin;";
                }
?>
<tr>
    <td class="el"><small><?php echo $file_col ?></small></td>
    <td class="fl"><small><?php echo date('Y-m-d H:i:s', $scripts[$i]['mtime'])?></small></td>
    <td class="fl"><small><?php echo date('Y-m-d H:i:s', $scripts[$i]['ctime'])?></small></td>
    <td class="fr"><small><?php echo $ttl_col ?></small></td>
    <td class="fr"><small><?php echo format_size($scripts[$i]['size'])?></small></td>
    <td class="fr"><small><?php echo number_format($scripts[$i]['hits'])?></small></td>
</tr>
<?php
      }
?>
<tr>
    <td colspan="1" style="text-align: left">&nbsp;</td>
    <td colspan="4" style="text-align: right;"><small><?php echo pageselstr($pg, ceil($numtot/$npp))?></small></td>
</tr>
</table>
<?php
            }
            break;
    }

print_footer();
?>
